File: app/resident_manager/lib/src/models/info.dart
import "snowflake.dart";

import "../utils.dart";

/// Data model for objects holding personal information.
class PersonalInfo {
  /// The target's full name.
  final String name;

  /// The target's room number.
  final int room;

  /// The target's date of birth.
  final Date? birthday;

  /// The target's phone number.
  final String phone;

  /// The target's email address.
  final String? email;

  /// Constructs a [PersonalInfo] object with the given [name], [room], [birthday], [phone], and [email].
  const PersonalInfo({
    required this.name,
    required this.room,
    this.birthday,
    required this.phone,
    this.email,
  });

  /// Convert this object into a JSON-encodable map.
  Map<String, dynamic> toJson() {
    return {
      "name": name,
      "room": room,
      "birthday": birthday?.format("yyyy-mm-dd"),
      "phone": phone,
      "email": email,
    };
  }

  /// Generates a map containing personal information query parameters.
  ///
  /// This method returns a [Map] where the keys and values are both [String]s.
  /// The keys represent the parameter names, and the values represent the
  /// corresponding parameter values.
  ///
  /// Returns:
  ///   A map containing the personal information query parameters.
  Map<String, String> personalInfoQuery() {
    return {
      "name": name,
      "room": room.toString(),
      if (birthday != null) "birthday": birthday!.format("yyyy-mm-dd"),
      "phone": phone,
      if (email != null) "email": email!,
    };
  }
}

/// Data model for objects holding personal info along with a snowflake ID.
class PublicInfo extends PersonalInfo with Snowflake {
  @override
  final int id;

  /// The username information, this data is not always available.
  String? username;

  /// The hashed password information, this data is not always available.
  String? hashedPassword;

  /// Constructs a [PublicInfo] object with the given [id], [name], [room], [birthday], [phone], and [email].
  PublicInfo({
    required this.id,
    required super.name,
    required super.room,
    super.birthday,
    required super.phone,
    super.email,
    this.username,
    this.hashedPassword,
  });

  /// Constructs a [PublicInfo] object from a JSON object.
  PublicInfo.fromJson(dynamic data)
      : this(
          id: data["id"] as int,
          name: data["name"] as String,
          room: data["room"] as int,
          birthday: data["birthday"] == null ? null : Date.parse(data["birthday"] as String),
          phone: data["phone"] as String,
          email: data["email"] as String?,
          username: data["username"] as String?,
          hashedPassword: data["hashed_password"] as String?,
        );
}

File: server/v1/models/info.py
from __future__ import annotations

from datetime import date
from typing import Annotated, Optional

import pydantic

from .results import Result
from .snowflake import Snowflake
from ...utils import (
    validate_name,
    validate_room,
    validate_phone,
    validate_email,
)


__all__ = ("PersonalInfo", "PublicInfo")


class PersonalInfo(pydantic.BaseModel):
    """Data model for objects holding personal information."""

    name: Annotated[str, pydantic.Field(description="The full name of the resident")]
    room: Annotated[int, pydantic.Field(description="The room number of the resident")]
    birthday: Annotated[Optional[date], pydantic.Field(description="The resident's date of birth")] = None
    phone: Annotated[str, pydantic.Field(description="The resident's phone number")]
    email: Annotated[Optional[str], pydantic.Field(description="The resident's email")] = None

    def to_personal_info(self) -> PersonalInfo:
        return PersonalInfo(
            name=self.name,
            room=self.room,
            birthday=self.birthday,
            phone=self.phone,
            email=self.email,
        )

    def validate_info(self) -> Optional[Result[None]]:
        if self.email is None or len(self.email) == 0:
            self.email = None

        if not validate_name(self.name):
            return Result(code=101, data=None)

        if not validate_room(self.room):
            return Result(code=102, data=None)

        if not validate_phone(self.phone):
            return Result(code=103, data=None)

        if self.email is not None and not validate_email(self.email):
            return Result(code=104, data=None)

        return None


class PublicInfo(Snowflake, PersonalInfo):
    """Data model for objects holding personal info along with a snowflake ID."""

    def to_public_info(self) -> PublicInfo:
        return PublicInfo(
            id=self.id,
            name=self.name,
            room=self.room,
            birthday=self.birthday,
            phone=self.phone,
            email=self.email,
        )
